funflow-examples/makefile-tool/src/Main.hs

```diff
@@ -0,0 +1,256 @@
+import Types
+
+type Set = Set.Set
+type Map = Map.Map
+type FileName = String
+type FileContent = String
+
+main :: IO ()
+main = do
+  perhapsMakeFile <- getValidMakeFile Nothing
+  case perhapsMakeFile of
+    Right (MFError errMsg) -> putStrLn $ "Invalid make file:\n" ++ errMsg
+    Left mfile -> do
+      cwd <- getCurrentDir
+      let contentStore = cwd </> [reldir|makefiletest/store|]
+          defGoalRule = defaultGoal mfile
+          runCfg = getRunConfigWithoutFile contentStore
+      putStrLn ("Attempting build:\n" ++ show defGoalRule)
+      result <- runFlowWithConfig runCfg (tryE @SomeException (buildTarget contentStore mfile defGoalRule)) () :: IO (Either SomeException (Content File))
+      putStrLn "Build attempt complete"
+      case result of
+        Left ex -> putStrLn $ "\n\nFailed, target failed:\n\n" ++ (show ex)
+        Right execFile -> do
+          -- Build succeeded; write executable to target location and set exec bit.
+          let outpath = fromAbsDir cwd ++ "/" ++ mkRuleTarNm defGoalRule
+          readByteString contentStore execFile >>= BS.writeFile outpath
+          setFileMode outpath accessModes -- chmod +x
+          putStrLn "\n\nSuccess, target executable made."
+
+
+-- | Building A Target
+--------------------------------------------------------------------------------
+-- Note: assuming the makefile is valid at this point!
+buildTarget :: Path Abs Dir -> MakeFile -> MakeRule -> Flow () (Content File)
+buildTarget storeRoot mkfile target@(MakeRule targetNm deps cmd) = let
+   srcfiles = sourceFiles mkfile
+   -- What must be built is the collection of non-source dependencies.
+   neededTargets = Set.toList $ Set.difference deps srcfiles
+   -- We only need the sources that are pertinent for the current rule's dependencies.
+   neededSources  = Set.toList $ deps `Set.intersection` srcfiles
+   maybeFindDepRules = findRules mkfile neededTargets
+ in case maybeFindDepRules of
+   Nothing -> failNow
+   Just (depRules :: [MakeRule]) -> let
+     -- found rules for dependencies to be built for current rule
+       grabSrcsActions = mapM (readFile . ("./" ++))
+     in proc _ -> do
+       msgFlow ("Current rule: " ++ show target) -< ()
+       () <- failGuardFlow -< (target `Set.member` (allrules mkfile))
+       -- Read content from source files.
+       contentSrcFiles <- (ioFlow grabSrcsActions) -< neededSources
+       -- Result from building each dependency.
+       depFiles <- flowJoin [Id{ unId = buildTarget storeRoot mkfile r } | r <- depRules] -< (replicate (length depRules) ())
+       let fullSrcFiles = Map.fromList $ zip neededSources contentSrcFiles
+       -- Compile the target of the current rule.
+       compFile <- (compileFile storeRoot) -< (targetNm, fullSrcFiles, depFiles, cmd)
+       returnA -< compFile
+
+
+-- | Compiles a C file in a docker container.
+compileFile :: Path Abs Dir -> Flow (TargetFile, Map.Map SourceFile String, [Content File], Command) (Content File)
```

Comment:
I noticed that in the README we claim that this example caches build steps and initially thought that we had forgotten to port that logic over. After looking at the original code from funflow, I realized that we were not caching the build steps in the original example either :sweat_smile:. Since this wasn't implemented in the original example, I'm happy merging this PR and updating the caching in a later PR as well. 

At first glance it looks like we might want to apply the caching wherever we call `compileFile` using `Control.Kernmantle.Caching.caching'` [link](https://tweag.github.io/funflow2/api/kernmantle-caching/html/Control-Kernmantle-Caching.html#v:caching-39-).

Reply:
Another example https://tweag.github.io/funflow2/tutorials/QuickReference.html#Caching-a-flow. I think we'll want to use caching' instead to avoid having to generate unique ids for each of the build steps manually.